docs/RFCS/order_and_parallelization_of_mutations.md

```diff
@@ -0,0 +1,204 @@
+- Feature Name: Reordering and parallelization of multi-row mutations
+- Status: draft
+- Start Date: 2018-03-12
+- Authors: knz
+- RFC PR: (PR # after acceptance of initial draft)
+- Cockroach Issue: #23698
+
+**Remember, you can submit a PR with your RFC before the text is
+complete. Refer to the [README](README.md#rfc-process) for details.**
+
+# Summary
+
+A short write-up of the restrictions to re-order and parallelize
+INSERT/DELETE/UPDATE/UPSERT (mutation statements) when they operate
+over multiple rows.
+
+tl;dr: the data source of a mutation can always be distributed. UPDATE
+and DELETE can be distributed too if there are no FK cascading
+actions. In the other cases, "it depends". The RFC below details these
+cases.
+
+# Motivation
+
+Optimization of queries that include mutation statements over multiple rows.
+
```

Comment:
I appreciate you writing this up, my biggest concern is adding in functionality that users have not asked for. While the project itself is interesting, it adds a lot of complexity to our implementation. Before we consider the implementation details is there enough consensus that we want to do this?

Reply:
This is not a RFC that drives a change in CockroachDb. This is a RFC that specifies constraints on any future work to do these changes. It's similar in scope to the RFC `20161215_join_filters.md` which specified how the work could be done and was merged long before the change was actually done.